src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1994,6 +2024,129 @@ IntersectPlacementList(List *lhsPlacementList, List *rhsPlacementList)
 
 
 /*
+ * GroupInsertValuesByShardId takes care of grouping the rows from a multi-row
+ * INSERT by target shard. At this point, all pruning has taken place and we
+ * need only to build sets of rows for each destination. This is done by a
+ * simple sort (by shard identifier) and gather step. The sort has the side-
+ * effect of getting things in ascending order to avoid unnecessary deadlocks
+ * during Task execution.
+ */
+static List *
+GroupInsertValuesByShardId(List *insertValuesList)
+{
+	ModifyRoute *route = NULL;
+	ListCell *insertValuesCell = NULL;
+	List *modifyRouteList = NIL;
+
+	insertValuesList = SortList(insertValuesList, CompareInsertValuesByShardId);
```

Comment:
Good/subtle catch. I didn't realize that users might intentionally craft the order of multi-row INSERTs to avoid unnecessary deadlocks. Adding a secondary sort key (list index) for now to add stability to the algorithm. Fixing in a better way tomorrow morning.